Add randomness to leader selection

linked0 opened the pull request:
By adding the `random seed` for all the validators in calculating the hash value of nodes, we can add additional randomness when doing leader selection for each round. This PR has the implementation of that algorithm.

Fixes #1079

linked0 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -577,9 +577,8 @@ public class EnrollmentManager
             const preimage = this.validator_set.getPreimageAt(key, height);
-            // this should not happen. validators which didn't reveal the
-            // preimage should not be in the active validator set 'keys'.
-            assert(preimage != PreImageInfo.init);
```
> assert(preimage != PreImageInfo.init);

This should be restored, which will be dealt with in another PR. Currently, this makes many assertion failures due to timing issues which means that the function is called before the right pre-images are not revealed in many test cases.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +563,61 @@ extern(D):
+        ref const(NodeID) node_id) nothrow
+    {
+        uint512 hash;
+        const random_seed = this.ledger.getValidatorRandomSeed(Height(slot_idx));
```
You're supposed to use the previous height, not the current one..

AndrejMitrovic replied:
Lets consider the case where block height 1 was just externalized. We will have at least the preimages for height 1, but there is no guarantee we have preimages for heights 2+.

So when we start nominating block height 2, we can only take the most recent guaranteed preimages - the ones for height 1 

So you just have to do `slot_idx - 1` in this call.

linked0 replied:
> So when we start nominating block height 2, we can only take the most recent guaranteed preimages - the ones for height 1

I got it. Thanks.

AndrejMitrovic commented on source/agora/common/SCPHash.d:
```diff
@@ -60,3 +61,11 @@ public uint512 getHashOf (uint64_t slot_idx, ref const(Value) prev, uint32_t has
+public uint512 getHashOf (uint64_t slot_idx, ref const(Value) prev, uint32_t hash,
+    int32_t round_num, ref const(NodeID) node_id, ref const(AgoraHash) random_seed)
+{
+    return uint512(hashMulti(slot_idx, prev[], hash, round_num, node_id,
```
There is no need for this function. You can move this entire call inline into `computeHashNode`

linked0 replied:
Yes I can, but It's overloaded from the `getHashOf` and it's not bad to show that it's overloaded for the new feature. Don't you agree with this?

AndrejMitrovic replied:
There is a difference. Those other overloads are used by C++ code. This function is not.

AndrejMitrovic commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -577,9 +577,8 @@ public class EnrollmentManager
         foreach (const ref key; keys)
         {
             const preimage = this.validator_set.getPreimageAt(key, height);
-            // this should not happen. validators which didn't reveal the
```
Please revert these changes, this assert is a preventative measure so the function is not called with the wrong height.

linked0 replied:
I'm trying to restore the removed assertion, `assert(preimage != PreImageInfo.init);` by adding the code into the test codes like `QuorumPreimage.d` as follows.

```
foreach (enroll; enrolls)
{
    nodes.each!(node => 
        retryFor(node.getPreimage(enroll.utxo_key).distance >= 1 , 10.seconds));
}
```

But, this code fails after outsider validators are enrolled. I guess it's because the outsider validators don't reveal their pre-images timely although they are enrolled as a validator. I need more debugging for that. Is there any variable to distinguish original validators from outsider validators in the code? Any clues would be very helpful. Thanks in advance.

linked0 replied:
```
foreach (enroll; blocks[0].header.enrollments)
{
    nodes.each!(node =>
        retryFor(node.getPreimage(enroll.utxo_key).distance >= 1 , 10.seconds));
}
```

With the codes like above, we can avoid assertions failure in many tests in `agora.test`. But, In QuorumPreimage.d, some failures occur continuously, which is related to outsiders I think. I’m debugging for that.

Geod24 replied:
~Yes, this needs to be reverted. I can take a look.~

TrustHenry commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +563,61 @@ extern(D):
+            is_priority = the flag to check that this call is for priority.
+            round_number = the nomination round
+            node_id = the id of the node for which this computation is being made
+
+    ***************************************************************************/
+
+    public override uint64_t computeHashNode(uint64_t slot_idx,
```
No document definition for the return value. `uint64_t`

linked0 replied:
No document definition for the return value. uint64_t

> I will add a description for that. Thanks.

TrustHenry commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +563,61 @@ extern(D):
+
+    ***************************************************************************/
+
+    static private uint64_t hashHelper(ref const(uint512) hash) nothrow
```
No document definition for the return value. `uint64_t`

AndrejMitrovic replied:
```suggestion
    private static uint64_t hashHelper (ref const(uint512) hash) nothrow
```

`private` goes first. And the `(` styling fix.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -507,6 +507,31 @@ public class Ledger
+
+    ***************************************************************************/
+
+    public Hash getValidatorRandomSeed (Height height) nothrow
```
This is awkward code, it does not belong here.

There is nothing this function uses in the Ledger itself other than the enrollment manager. 

Why not move `this.enroll_man.getEnrolledUTXOs(keys)` inside of `EnrollmentManager.getRandomSeed` and remove the `in Hash[] keys` parameter. And then also provide the `EnrollmentManager` instance to the `Nominator` and make it call `getRandomSeed(slot_idx - 1)`?

linked0 replied:
> There is nothing this function uses in the Ledger itself other than the enrollment manager.

All right. I have already felt something weird about that and you have suggested the right approach. I'll apply it to the code. Thanks.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +563,61 @@ extern(D):
+
+    /***************************************************************************
+
+        Convert the `uint512` type of hash value into the `uint64_t` type
+
+        Params:
+            hash = the hash value of `uint512` type.
```
For very simple functions we can just use a `Params` and `Returns` section if the function returns something.

```suggestion
        Params:
            hash = the hash value of `uint512` type.

        Returns:
            An 8-byte `uint64_t` value reduced from the 64-byte `uint512` hash value
```

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +563,61 @@ extern(D):
+        {
+            log.fatal("Computing hash of the node({}) failed: {}, Data was: ",
+                node_id, ex.msg);
+            log.fatal("\tslot_idx: {}, prev: {}, is_priority: {}, randon_seed: {}",
```
```suggestion
            log.fatal("\tslot_idx: {}, prev: {}, is_priority: {}, random_seed: {}",
```

Why isn't this all in one log call? You can add `\n` if you really want to add a newline.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -75,6 +76,12 @@ public extern (C++) class Nominator : SCPDriver
 
+    // values used to switch hash function between priority and neighborhood checks
+    // which are from SCPDriver.cpp
+    const uint hash_N = 1;
+    const uint hash_P = 2;
+    const uint hash_K = 3;
```
Please move these inside of `computeHashNode`. Constants should only be in an outer scope if they're used by other code too, but they are not used anywhere else.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +563,61 @@ extern(D):
+
+    ***************************************************************************/
+
+    public override uint64_t computeHashNode(uint64_t slot_idx,
```
```suggestion
    public override uint64_t computeHashNode (uint64_t slot_idx,
```

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +563,61 @@ extern(D):
+            assert(0);
+        }
+
+        return hashHelper(hash);
```
Copy-paste the implementation of `hashHelper` inline here, because it's not used anywhere else.

AndrejMitrovic reviewed (changes requested):
Few changes requested.

linked0 committed:
Add randomness to leader selection

By adding the `random seed` for all the validators in calculating the hash value
of nodes, we can add additional randomness when doing leader selection for each
round. This commit implements that algorithm.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +557,53 @@ extern(D):
+            this.ledger.getValidatorRandomSeed(Height(slot_idx - 1));
+        try
+        {
+            hash = uint512(hashMulti(slot_idx, prev[], hash, round_num, node_id,
```
```suggestion
            hash = uint512(hashMulti(slot_idx, prev[], isPriority ? hash_P : hash_N, hash, round_num, node_id,
```

AndrejMitrovic replied:
Ping.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -556,6 +557,53 @@ extern(D):
+        const uint hash_N = 1;
+        const uint hash_P = 2;
+
+        uint512 hash;
+        const random_seed =
+            this.ledger.getValidatorRandomSeed(Height(slot_idx - 1));
```
```suggestion
        const random_seed =
            this.ledger.getValidatorRandomSeed(Height(slot_idx - 1));
        uint512 hash;
```

The declaration should be moved the closest to where it's initialized.

AndrejMitrovic replied:
Ping.

linked0 committed:
Make `expectBlock` check the needed pre-images are revealed

Sometimes we need to check the needed pre-images have been already revealed
before making sure that the nodes generates a block at the given block.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -518,6 +518,48 @@ public class Ledger
+
+    ***************************************************************************/
+
+    public bool hasTransactionHash (const ref Hash tx) @safe
```
Why did you reintroduce this? it was removed in a previous PR.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -518,6 +518,48 @@ public class Ledger
+    ***************************************************************************/
+
+    public Hash getValidatorRandomSeed (Height height) nothrow
+    {
```
Why did you reintroduce it again?

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -553,6 +553,23 @@ public class TestAPIManager
                     idx, node.getBlockHeight(), height), file, line));
     }
 
+    public void expectBlock (Height height, const(BlockHeader) enrolled_header,
```
This needs documentation. Actually if it only checks the enrollments then it might be better to just pass `b0.header.enrollments` instead as the parameter.

linked0 replied:
In fact, it needs `height` because the `waitForPreimages` in the function is called with a distance which is the requested height minus the enrolled height. I documented that.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -553,6 +553,23 @@ public class TestAPIManager
+            line);
+    }
+
+    public void expectBlock (Clients)(Clients clients, Height height,
```
Needs a `/// Ditto`

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -707,6 +708,53 @@ extern(D):
+            node_id = the id of the node for which this computation is being made
+
+        Returns:
+            the 64-byte hash which is unsigned long long type
```
```suggestion
            the 8-byte hash
```

It's 64-bit (8-byte), not 64-byte. And `unsigned long long` is C++ terminology, we should not use it here.

linked0 commented:
After rebasing, `SIGABRT` suddenly occurs in the unit test of agora.test.EnrollmentManager as follows. It occurs during calling the `rollback` of the `ManagedDatabase` in `addValidatedBlock` function.


```
(lldb) r
Process 50685 launched: '/Users/jaylee/work/agora/build/agora-unittests' (x86_64)
Process 50685 stopped
* thread #4, stop reason = signal SIGABRT
    frame #0: 0x00007fff6a52d33a libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff6a52d33a <+10>: jae    0x7fff6a52d344            ; <+20>
    0x7fff6a52d33c <+12>: movq   %rax, %rdi
    0x7fff6a52d33f <+15>: jmp    0x7fff6a527629            ; cerror_nocancel
    0x7fff6a52d344 <+20>: retq
Target 0: (agora-unittests) stopped.
(lldb) bt
error: need to add support for DW_TAG_base_type 'char' encoded with DW_ATE = 0x10, bit_size = 8
error: need to add support for DW_TAG_base_type 'immutable(char)' encoded with DW_ATE = 0x10, bit_size = 8
* thread #4, stop reason = signal SIGABRT
  * frame #0: 0x00007fff6a52d33a libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff6a5e9e60 libsystem_pthread.dylib`pthread_kill + 430
    frame #2: 0x00007fff6a4b4808 libsystem_c.dylib`abort + 120
    frame #3: 0x00000001001bdbbb agora-unittests`_D5agora6common15ManagedDatabaseQr8rollbackFNbNeZv at ManagedDatabase.d:165:17
    frame #4: 0x00000001001bf590 agora-unittests`_D5agora4node6LedgerQh17addValidatedBlockMFNfKxSQBu9consensus4data5BlockQgZv(this=0x00000001b5d0d5c0, block=0x00000002b6617590) at Ledger.d:290:25
    frame #5: 0x00000001001bf2fb agora-unittests`_D5agora4node6LedgerQh11acceptBlockMFNfKxSQBo9consensus4data5BlockQgZb(this=0x00000001b5d0d5c0, block=0x00000002b6617590) at Ledger.d:233:9
    frame #6: 0x00000001001befe1 agora-unittests`_D5agora4node6LedgerQh14onExternalizedMFNeSQBp9consensus4data13ConsensusDataQpZb(this=0x00000001b5d0d5c0, data=ConsensusData @ 0x00000002b66176b0) at Ledger.d:208:9
    frame #7: 0x0000000100154193 agora-unittests`Nominator::valueExternalized(this=0x00000001b5cbb160, slot_idx=1, value=<unavailable>) at Nominator.d:579:13
    frame #8: 0x0000000100630186 agora-unittests`stellar::BallotProtocol::setConfirmCommit(this=0x00000001044c9548, c=0x00000002b6617a40, h=0x00000002b6617a20) at BallotProtocol.cpp:1517:26
    frame #9: 0x000000010062fc29 agora-unittests`stellar::BallotProtocol::attemptConfirmCommit(this=0x00000001044c9548, hint=0x00000002b66190c8) at BallotProtocol.cpp:1493:16
    frame #10: 0x00000001006277e4 agora-unittests`stellar::BallotProtocol::advanceSlot(this=0x00000001044c9548, hint=0x00000002b66190c8) at BallotProtocol.cpp:1877:15
    frame #11: 0x00000001006260e7 agora-unittests`stellar::BallotProtocol::processEnvelope(this=0x00000001044c9548, envelope=0x00000002b66190c8, self=true) at BallotProtocol.cpp:199:13
    frame #12: 0x00000001006e1bb1 agora-unittests`stellar::Slot::processEnvelope(this=0x00000001044c9528, envelope=0x00000002b66190c8, self=true) at Slot.cpp:143:35
    frame #13: 0x0000000100628ccf agora-unittests`stellar::BallotProtocol::emitCurrentStateStatement(this=0x00000001044c9548) at BallotProtocol.cpp:626:19
    frame #14: 0x000000010062c8d6 agora-unittests`stellar::BallotProtocol::setPreparedAccept(this=0x00000001044c9548, ballot=0x00000002b6619590) at BallotProtocol.cpp:917:9
    frame #15: 0x000000010062be1f agora-unittests`stellar::BallotProtocol::attemptPreparedAccept(this=0x00000001044c9548, hint=0x00000002b661ad30) at BallotProtocol.cpp:881:20
    frame #16: 0x0000000100627736 agora-unittests`stellar::BallotProtocol::advanceSlot(this=0x00000001044c9548, hint=0x00000002b661ad30) at BallotProtocol.cpp:1871:15
    frame #17: 0x00000001006260e7 agora-unittests`stellar::BallotProtocol::processEnvelope(this=0x00000001044c9548, envelope=0x00000002b661ad30, self=false) at BallotProtocol.cpp:199:13
    frame #18: 0x00000001006e1bb1 agora-unittests`stellar::Slot::processEnvelope(this=0x00000001044c9528, envelope=0x00000002b661ad30, self=false) at Slot.cpp:143:35
    frame #19: 0x00000001006d0f80 agora-unittests`stellar::SCP::receiveEnvelope(this=0x00000001064a8e80, envelope=0x00000002b661ad30) at SCP.cpp:33:38
    frame #20: 0x0000000100154c33 agora-unittests`_D5agora9consensus8protocol9NominatorQk15receiveEnvelopeMFNeMKxS4scpd5types11Stellar_SCP11SCPEnvelopeZv(this=0x00000001b5cbb160, envelope=0x00000002b661ad30) at Nominator.d:466:9
    frame #21: 0x0000000100155a1e agora-unittests`_D5agora4node9ValidatorQk15receiveEnvelopeMFNfS4scpd5types11Stellar_SCP11SCPEnvelopeZv(this=0x0000700011813700, envelope=SCPEnvelope @ 0x00000002b661ad30) at Validator.d:260:9
    frame #22: 0x000000010017ec6f agora-unittests`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPITSQBaQxQu10SerializerZQCg13handleCommandFSQDuQDq7CommandQDdSQEmQEi9FilterAPIZv(cmd=Command @ 0x00000002b661bec0, node=0x0000700011813960, filter=FilterAPI @ 0x00000002b661bea0) at LocalRest.d-mixin-477:494:33
    frame #23: 0x00000001001a9c03 agora-unittests`_D6geod249LocalRest__T9RemoteAPITC5agora4test4Base7TestAPITSQBaQxQu10SerializerZQCg__T7spawnedTCQCkQCh17EnrollmentManager18__unittest_L287_C1FZ20MisbehavingValidatorZQDcFNbSQGp11concurrency3TidAyaiSQGi6common6ConfigQhPOSQIk8RegistryQjAySQHv9consensus4data5BlockQgmPOlPOmZ7runNodeMFZ__T6handleTSQLgQLc7CommandZQyMFQuZ9__lambda2MFZv at LocalRest.d:602:43
    frame #24: 0x0000000100787b19 agora-unittests`_D6geod2411concurrency14FiberScheduler6createMFNbDFZvZ4wrapMFZv at concurrency.d:916:13
    frame #25: 0x0000000100c5a36f agora-unittests`_D4core6thread7context8Callable6opCallMFZv(this=0x0000000000000002) at context.d:46:17
    frame #26: 0x0000000100c5a4fe agora-unittests`_D4core6thread5fiber5Fiber3runMFZv(this=0x0000000279240880) at fiber.d:1127:9
    frame #27: 0x0000000100c5a433 agora-unittests`fiber_entryPoint at fiber.d:184:13
```


In fact, [this line](https://github.com/bpfkorea/agora/blob/320a7981ca93eaaee4bf067cd16b78db603a1041/source/agora/test/EnrollmentManager.d#L416) makes the `SIGABRT`.

Will analyze more.

linked0 commented:
Ready for review.
I solved the above `SIGABRT` issue by calling the `waitForPreimage` function.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -564,6 +567,25 @@ public class TestAPIManager
+        const(BlockHeader) enrolled_header, Duration timeout,
+        string file = __FILE__, int line = __LINE__) if (isInputRange!Clients)
+    {
+        auto distance = cast(ushort)(height - enrolled_header.height - 1);
```
Could you add `assert(height > enrolled_header.height)`? Otherwise this will overflow and turn into `ulong.max`.

linked0 replied:
> Could you add assert(height > enrolled_header.height)? Otherwise this will overflow and turn into ulong.max.

Yes. I will. I think that's needed.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -518,6 +518,48 @@ public class Ledger
+
+    public bool hasTransactionHash (const ref Hash tx) @safe
+    {
+        return this.pool.hasTransactionHash(tx);
```
@linked0 Ping. This forwarding function does not need to exist.

linked0 replied:
> This forwarding function does not need to exist.

I'll check it. Ty.

AndrejMitrovic commented:
LGTM. Just please remove that `hasTransactionHash` wrapper. I think it was left in accidentally.

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -17,8 +17,9 @@ import agora.common.crypto.ECC;
 import agora.common.crypto.Schnorr;
 import agora.common.Config;
-import agora.common.Hash : Hash, HashDg, hashPart, hashFull;
+import agora.common.Hash : Hash, HashDg, hashPart, hashFull, hashMulti;
```
At this point, just import the whole module

linked0 replied:
> At this point, just import the whole module

If I import the whole module, the compiler outputs error as follows.
```
source/agora/consensus/protocol/Nominator.d(84,35): Error: alias agora.common.Hash.Hash at source/agora/common/Hash.d(86,8) conflicts with alias scpd.types.Stellar_types.Hash at source/scpd/types/Stellar_types.d(32,1) 
```

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -543,8 +543,9 @@ public class EnrollmentManager
-            // that is needed.
+            // this should not happen. validators which didn't reveal the
+            // preimage should not be in the active validator set 'keys'.
+            assert(preimage != PreImageInfo.init);
```
Looks like  the second commit should come first, so the assertion doesn't need to be deleted in the first place ?

linked0 replied:
Yeah, it would be good to come first.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -540,6 +540,9 @@ public class TestAPIManager
         for all nodes (this is what most tests expect).
 
+        Other overloads allow passing a `BlockHeader` of an enrolled height to
+        get enrollments and a height for checking pre-images are revealed.
```
This seems to me like an abuse of overloading. The other overloads do much more (checking pre-images) than this. Perhaps the other overloads should just be the default ? Although the client doesn't care much about pre-images.

linked0 replied:
I'm thinking about what the best document for that is. I'll add the right description.

Geod24 reviewed (commented):
The second commit looks like a workaround for a deeper problem, but if it allows to get this moved forward, I think it's okay.
Just invert the commits, so the removal of the assertion doesn't need to happen.

linked0 commented:
Ready for review. Thanks for your reviews.
